EDF conversion fixes

tab-cmd opened the pull request:
# Overview

Updates to improve EDF data conversions. Applying trigger offsets, ability to write targetness, and handling cases where there are too many annotations. 

## Ticket

- https://www.pivotaltracker.com/story/show/176668858
- https://www.pivotaltracker.com/story/show/176650169

## Contributions

- `convert.py`: Add mode, write_targetness, and annotation_channels keyword arguments.
- `demo_convert.py`: Update to add new keywords
- `helpers/load.py`: add extract_mode method for determining the mode
- `helpers/triggers.py`: add an offset correction method
- `helpers/vizualization.py`: moved plot_edf function from demo to this module.
- `task_registry.py`: Update tasks to include the necessary mode

## Test

- Added unittest
- Converted several files and had @danielpklee confirm in BrainVision

## Documentation

- Are documentation updates required? In-line, README, or [documentation](https://github.com/BciPy/bcipy.github.io)? Yes. I added task/mode to the README dictionary.

tab-cmd committed:
EDF conversion fixes 




Update demo code


Add tests and cleanup


lint

lawhead commented on bcipy/helpers/convert.py:
```diff
@@ -80,12 +124,15 @@ def write_edf(output_path: str,
                            n_channels=n_channels,
                            file_type=FILETYPE_EDFPLUS)
+        if annotation_channels:
+            writer.set_number_of_annotation_signals(annotation_channels)
```
I read through the documentation for this method and it seems like this is needed any time the number of triggers/annotations exceeds the number of seconds in the experiment. Wouldn't this almost always be the case since we usually show about 3 to 4 stim per second? Should we try to calculate this from the parameters (using stim_length)?

tab-cmd replied:
In practice this wasn't clear cut and forcing the annotation channels led to some issues. I added a warning method to let the user know some triggers may be lost and to increment gradually.

lawhead commented on bcipy/helpers/convert.py:
```diff
@@ -41,19 +55,45 @@ def convert_to_edf(data_dir: str,
+    return write_edf(edf_path, raw_data, ch_names, sfreq, events, overwrite, annotation_channels)
+
+
+def compile_triggers(labels, targetness, timing, write_targetness):
```
This method could use a docstring.

lawhead commented on bcipy/helpers/convert.py:
```diff
@@ -8,16 +8,23 @@
                    edf_path: str = None,
                    overwrite=False,
-                   use_event_durations=False) -> Path:
+                   write_targetness=True,
```
Do you think it's a more common scenario that we would want to see the targetness rather than the stimuli? Should the default be False?

lawhead commented on bcipy/helpers/tests/test_convert.py:
```diff
@@ -92,25 +97,63 @@ def test_convert_defaults(self):
+        """Test creating the EDF without event annotations"""
+        path = convert_to_edf(self.temp_dir,
+                              mode=self.default_mode,
+                              write_targetness=True,
```
If this is the default we probably want a separate test where `write_targetness=False` to check that the stim name is written correctly.

lawhead reviewed (approved):
Nice work figuring this out. This one was tricky!

tab-cmd committed:
Address PR feedback

add warning for trigger_count > recording time, add documentation, use False for write targetness default, update variable names, lint

tab-cmd committed:
fix typo
